vnext/ReactUWP/Utils/BaseScriptStoreImpl.cpp

```diff
@@ -0,0 +1,341 @@
+
+#include <codecvt>
+#include <locale>
+
+using namespace facebook;
+
+namespace winrt {
+using namespace winrt::Windows::Foundation;
+using namespace winrt::Windows::Storage;
+}; // namespace winrt
+
+namespace react {
+namespace uwp {
+
+namespace {
+
+class ByteArrayBuffer : public facebook::jsi::Buffer
+{
+public:
+  size_t size() const override
+  {
+    return size_;
+  }
+
+  const uint8_t* data() const
+  {
+    return byteArray_.get();
+  }
+
+  uint8_t* data()
+  {
+    return byteArray_.get();
+  }
+
+  ByteArrayBuffer(size_t size)
+    : size_(size), byteArray_(std::make_unique<uint8_t[]>(size)) {}
+
+  ByteArrayBuffer(ByteArrayBuffer&&) = default;
+  ByteArrayBuffer& operator=(ByteArrayBuffer&&) = default;
+
+private:
+  ByteArrayBuffer(const ByteArrayBuffer&) = delete;
+  ByteArrayBuffer& operator=(const ByteArrayBuffer&) = delete;
+
+  size_t size_;
+  std::unique_ptr<uint8_t[]> byteArray_;
+};
+
+class BufferViewBuffer : public facebook::jsi::Buffer
+{
+public:
+  size_t size() const override
+  {
+    return size_;
+  }
+
+  const uint8_t* data() const
+  {
+    return buffer_->data() + offset_;
+  }
+
+  uint8_t* data()
+  {
+    return const_cast<uint8_t*>(buffer_->data()) + offset_;
+  }
+
+  BufferViewBuffer(std::unique_ptr<const facebook::jsi::Buffer> buffer, size_t offset, size_t size)
+    : buffer_(std::move(buffer)), offset_(offset), size_(size)
+  {
+    if (size_ > buffer_->size() - offset)
+      std::terminate();
+  }
+
+  BufferViewBuffer(BufferViewBuffer&&) = default;
+  BufferViewBuffer& operator=(BufferViewBuffer&&) = default;
+
+private:
+  BufferViewBuffer(const BufferViewBuffer&) = delete;
+  BufferViewBuffer& operator=(const BufferViewBuffer&) = delete;
+
+  std::unique_ptr<const facebook::jsi::Buffer> buffer_;
+  size_t offset_;
+  size_t size_;
+};
+
+constexpr const char* PERSIST_MAGIC = "CHAKRAPREP";
+constexpr const char* PERSIST_EOF = "EOF";
+
+int constexpr length__(const char* str)
+{
+  return *str ? 1 + length__(str + 1) : 0;
+}
+
+// TODO : Enforce any packing policy ?
+struct PreparedScriptPrefix 
+{
+  // TODO :: constexpr initialize the array.
+  char magic[length__(PERSIST_MAGIC)];
+  jsi::ScriptVersion_t scriptVersion;
+  jsi::JSRuntimeVersion_t runtimeVersion;
+  uint64_t sizeInBytes;
+};
+
+struct PreparedScriptSuffix 
+{
+  char eof[length__(PERSIST_EOF)];
+};
+
+}
+
+jsi::VersionedBuffer BaseScriptStoreImpl::getVersionedScript(const std::string& url) noexcept 
+{
+  std::ifstream file(url, std::ios::binary | std::ios::ate);
+
+  if (!file)
+  {
+    return { nullptr, 0 };
+  }
+
+  std::streamsize size = file.tellg();
+  file.seekg(0, std::ios::beg);
+
+  auto buffer = std::make_unique<ByteArrayBuffer>(static_cast<size_t>(size));
+  if (!file.read(reinterpret_cast<char*>(buffer->data()), size))
+  {
+    return { nullptr, 0 };
+  }
+
+  file.close();
+
+  return { std::move(buffer), versionProvider_? versionProvider_->getVersion(url) : static_cast<uint64_t>(size) };
```

Comment:
I think it is worth calling out that without a versionProvider_ this assumes that a larger file is a newer file. Should be mostly true, but refactoring or bundle cleanup can lead to newer smaller code.

Reply:
Thanks. Currently, we are only using preparedscriptstore, not script store with V8.. We are currently manually deleting the cache when we want to regenerate cache, which is terrible :) but it is enough for us to start performance comparisons … We will soon fix the runtime to use script stores and take the version into consideration while generating caches.. 
Even if some one uses the script store with BaseScriptStoreImpl, they are expected to instantiate BaseScriptStoreImpl with a ScriptVersionProvider instnace which knows the version of all the scripts out there .. Using size as a version is only a last resort fallback which we should never include with the shipped product. It should just treat the size as the version, and any change in size should be a signal that the script changed. Whether it looks for larger/smaller/different depends on how our runtime implementations uses the result. They should just check whether the version is different, not whether the version is newer when regenerating the bytecodes.
I prefer fixing the versioning as a separate change.

Reply:
sounds good. The part about versions changing at all vs versions mismatching is what I was missing. you're correct for this use case if they don't match at all they are not usable.